source/extensions/filters/http/fault/config.cc

```diff
@@ -44,7 +44,7 @@ FaultFilterFactory::createFilterFactoryFromProto(const Protobuf::Message& proto_
 
 Router::RouteSpecificFilterConfigConstSharedPtr
 FaultFilterFactory::createRouteSpecificFilterConfig(const Protobuf::Message& proto_config) {
-  return std::make_shared<const Router::RouteSpecificFilterConfig>(Fault::FaultSettings(
+  return std::shared_ptr<const Router::RouteSpecificFilterConfig>(new Fault::FaultSettings(
```

Comment:
With a bit of help from @jsedgwick, I think I've figured out a fix that doesn't require these changes to the base class: instead of `make_shared<const Router::RouteSpecificFilterConfig>`, use `make_shared<const FaultSettings>`. Integration tests and a unit test specifically covering this case passed locally in https://github.com/envoyproxy/envoy/pull/3120, just waiting on CI to confirm.

Reply:
We were constructing the base class directly from a temporary derived value (`make_shared<Base>(Derived())`) and therefore slicing (equivalently, `Base b = Derived();`). There is no longer a `Derived` anywhere so a later `dynamic_cast<Derived>` will fail.

Making the base class pure was a little bit of a red herring that broke the ability to construct the base, which is not what we wanted to do in the first place. `shared_ptr<Base>(new Derived())` converts the `Derived` *pointer* to a `Base`, which is what we want, but it's a little cleaner to just `make_shared<Derived>` and let the compiler implicitly convert it to a `shared_ptr<Base>` upon returning from this function.

Reply:
👀